tests/st/policy/test_felix_gateway.py

```diff
@@ -0,0 +1,510 @@
+                                "--cluster-store-opt kv.cacertfile=%s " \
+                                "--cluster-store-opt kv.certfile=%s " \
+                                "--cluster-store-opt kv.keyfile=%s " % \
+                                (ETCD_HOSTNAME_SSL, ETCD_CA, ETCD_CERT,
+                                 ETCD_KEY)
+else:
+    ADDITIONAL_DOCKER_OPTIONS = "--cluster-store=etcd://%s:2379 " % \
+                                get_ip()
+
+
+class TestFelixOnGateway(TestBase):
+    """
+    Tests that policy is correctly implemented when using Calico
+    on a gateway or router.  In that scenario, Calico should
+    police forwarded (possibly NATted) traffic using the host endpoint
+    policy.
+    """
+    hosts = None
+    gateway = None
+    host = None
+
+    @classmethod
+    def setUpClass(cls):
+        # Wipe etcd once before any test in this class runs.
+        wipe_etcd()
+
+        # We set up an additional docker network to act as the external
+        # network.  The Gateway container is connected to both networks.
+        # and we configure it as a NAT gateway.
+        #
+        #  "cali-st-ext" host
+        #   container
+        #      |
+        #  "cali-st-ext" docker
+        #    bridge
+        #      |
+        #  Gateway           Host
+        #  container         container
+        #         \          /
+        #        default docker
+        #            bridge
+
+        # First, create the hosts and the gateway.
+        cls.hosts = []
+        log_and_run("docker rm -f cali-st-gw || true")
+        cls.gateway = DockerHost("cali-st-gw",
+                                 additional_docker_options=ADDITIONAL_DOCKER_OPTIONS,
+                                 post_docker_commands=POST_DOCKER_COMMANDS,
+                                 start_calico=False)
+        cls.gateway_hostname = cls.gateway.execute("hostname")
+        log_and_run("docker rm -f cali-st-host || true")
+        cls.host = DockerHost("cali-st-host",
+                              additional_docker_options=ADDITIONAL_DOCKER_OPTIONS,
+                              post_docker_commands=POST_DOCKER_COMMANDS,
+                              start_calico=False)
+        cls.host_hostname = cls.host.execute("hostname")
+        cls.hosts.append(cls.gateway)
+        cls.hosts.append(cls.host)
+
+        # Delete the nginx container if it still exists.  We need to do this
+        # before we try to remove the network.
+        log_and_run("docker rm -f cali-st-ext-nginx || true")
+
+        # Create the external network.
+        log_and_run("docker network rm cali-st-ext || true")
+        # Use 172.19.0.0 to avoid clash with normal docker subnet and
+        # docker-in-docker subnet
+        log_and_run("docker network create --driver bridge --subnet 172.19.0.0/16 cali-st-ext")
+
+        # And an nginx server on the external network only.
+        log_and_run("docker run"
```

Comment:
If you use a 'workload' image, that has TCP and UDP servers built in (they'll echo back to you what you send them).  And it'll respond to ping too.